akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,377 @@
+        // also, ensure that there is an event handler for each single event
+        state = applyEvent(state, event)
+        val tags = tagger(event)
+        val eventToPersist = if (tags.isEmpty) event else Tagged(event, tags)
+
+        log.info(s" >>>> [${context.self.path.name}] in Persist($event) @ ${lastSequenceNr}")
+
+        internalPersist(eventToPersist) { _ ⇒
+          applySideEffects(sideEffects)
+
+          log.info(s" >>>> [${context.self.path.name}] in Persist($event) CALLBACK @ ${lastSequenceNr}")
+
+          log.info(s"persist callback [${context.self.path.name}] == seqNr=${lastSequenceNr} => ${snapshotWhen(state, event, lastSequenceNr)}")
+          if (snapshotWhen(state, event, lastSequenceNr))
+            internalSaveSnapshot(state)
+        }
+
+      case PersistAll(events) ⇒
+        if (events.nonEmpty) {
+          // apply the event before persist so that validation exception is handled before persisting
+          // the invalid event, in case such validation is implemented in the event handler.
+          // also, ensure that there is an event handler for each single event
+          var count = events.size
+          var seqNr = lastSequenceNr
+          val (newState, shouldSnapshotAfterPersist) = events.foldLeft((state, false)) {
+            case ((currentState, snapshot), event) ⇒
+              seqNr += 1
+              val shouldSnapshot = snapshot || snapshotWhen(currentState, event, seqNr)
+              (applyEvent(currentState, event), shouldSnapshot)
+          }
+          state = newState
+          val eventsToPersist = events.map { event ⇒
+            val tags = tagger(event)
+            if (tags.isEmpty) event else Tagged(event, tags)
+          }
+          internalPersistAll(eventsToPersist) { _ ⇒
+            count -= 1
+            if (count == 0) {
+              sideEffects.foreach(applySideEffect)
+              if (shouldSnapshotAfterPersist)
+                internalSaveSnapshot(state)
+            }
+          }
+        } else {
+          // run side-effects even when no events are emitted
+          sideEffects.foreach(applySideEffect)
+          same
+        }
+
+      case _: PersistNothing.type @unchecked ⇒
+        // ignore, no effects to apply
+        same
+
+      case _: Unhandled.type @unchecked ⇒
+        Behavior.unhandled
+
+      case c: ChainableEffect[_, S] ⇒
+        applySideEffect(c)
+    }
+  }
+
+  private def shouldNotHappen(): Behavior[Any] = {
+    throw new RuntimeException("Received message which should not happen in Running state!")
+  }
+
+  private def popApplyHandler(payload: Any): Unit =
+    try pendingInvocations.pop().handler(payload)
+    finally flushBatch()
+
+  private def flushBatch() {
+    if (eventBatch.nonEmpty) {
+      journalBatch ++= eventBatch.reverse
+      eventBatch = Nil
+    }
+
+    flushJournalBatch()
+  }
+
+  private def flushJournalBatch(): Behavior[Any] = {
+    if (!writeInProgress && journalBatch.nonEmpty) {
```

Comment:
it should be possible to simplify this batching and flushing stuff, since no async?

Reply:
Yes, removed all the batches :)